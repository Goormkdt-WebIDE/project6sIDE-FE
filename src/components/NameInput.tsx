import React from "react";
import { UseFormRegister, UseFormReturn } from "react-hook-form";

interface FormValue {
  name: string;
  email: string;
  password: string;
  password_confirm: string;
}

interface NameInputProps {
  register: UseFormRegister<FormValue>;
  errors: UseFormReturn<FormValue>["formState"]["errors"];
}

const NameInput: React.FC<NameInputProps> = ({ register, errors }) => {
  return (
    <>
      <input
        placeholder="Name"
        {...register("name", { required: true, maxLength: 10 })}
        className="border rounded-md p-2 w-full mt-4"
      />
      {errors.name && (
        <p className="text-red-500">
          <span className="inline-block align-middle mb-2.5">âš  </span>
          This name field is required
        </p>
      )}
      {errors.name?.type === "maxLength" && (
        <p className="text-red-500">Your input exceeds the maximum length</p>
      )}
    </>
  );
};

export default NameInput;
