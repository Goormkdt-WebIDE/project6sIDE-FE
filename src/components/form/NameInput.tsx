import React from "react";
import {
  FieldValues,
  Path,
  UseFormRegister,
  UseFormReturn,
} from "react-hook-form";

type Props<T extends FieldValues> = {
  register: UseFormRegister<T>;
  errors: UseFormReturn<T>["formState"]["errors"];
  name: keyof T;
  placeholder?: string;
};

function NameInput<T extends FieldValues>({
  register,
  errors,
  name,
  placeholder,
}: Props<T>) {
  return (
    <>
      <input
        placeholder={placeholder ? placeholder : "Name"}
        {...register(name as Path<T>, { required: true, maxLength: 10 })}
        className="border rounded-md p-2 w-full mt-4"
      />
      {errors[name] && (
        <p className="text-red-500" role="alert">
          <span className="inline-block align-middle mb-2.5">âš  </span>
          This name field is required
        </p>
      )}
      {errors[name]?.type === "maxLength" && (
        <p className="text-red-500">Your input exceeds the maximum length</p>
      )}
    </>
  );
}

export default NameInput;
