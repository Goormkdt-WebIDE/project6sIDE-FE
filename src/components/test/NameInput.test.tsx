import "@testing-library/jest-dom";
import { FieldError } from "react-hook-form";
import NameInput from "../NameInput";
import { render, screen } from "@testing-library/react";
import React from "react";

describe("NameInput", () => {
  const register = jest.fn();

  afterEach(() => {
    register.mockReset();
  });

  it("스냅샷 테스트", () => {
    const { asFragment } = renderNameInput();
    expect(asFragment()).toMatchSnapshot();
  });

  it("placeholder prop을 전달하면 그것을 올바르게 render 해야한다", () => {
    const placeholder = "Name을 입력하세요";
    renderNameInput(placeholder);
    expect(screen.getByPlaceholderText(placeholder)).toBeInTheDocument();
  });

  it("에러가 발생하면 밑에 에러가 표시되어야한다", () => {
    renderNameInput("", {} as FieldError);
    expect(screen.getByRole("alert")).toBeInTheDocument();
  });

  it("Max Length 에러가 발생하면 밑에 에러가 표시되어야한다", () => {
    renderNameInput("", { type: "maxLength" } as FieldError);
    expect(screen.getByRole("alert")).toBeInTheDocument();
  });

  function renderNameInput(placeholder?: string, error?: FieldError) {
    return render(
      <NameInput
        register={register}
        errors={{ name: error }}
        placeholder={placeholder}
      />
    );
  }
});
