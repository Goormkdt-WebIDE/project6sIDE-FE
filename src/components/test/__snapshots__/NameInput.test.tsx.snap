// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NameInput 스냅샷 테스트 1`] = `
<DocumentFragment>
  <input
    class="border rounded-md p-2 w-full mt-4"
    placeholder="Name"
  />
</DocumentFragment>
`;
